crates/ironrdp-dvc/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(unused)] // FIXME: remove this annotation
+
+// TODO: this crate is WIP
+
+#[macro_use]
+extern crate tracing;
+
+extern crate alloc;
+
+// Re-export ironrdp_pdu crate for convenience
+#[rustfmt::skip] // do not re-order this pub use
+pub use ironrdp_pdu as pdu;
+
+use alloc::borrow::ToOwned;
+use alloc::boxed::Box;
+use alloc::collections::BTreeMap;
+use alloc::string::String;
+use alloc::vec::Vec;
+use core::any::Any;
+use core::fmt;
+
+use ironrdp_pdu::gcc::ChannelName;
+use ironrdp_pdu::rdp::vc;
+use ironrdp_pdu::write_buf::WriteBuf;
+use ironrdp_pdu::{assert_obj_safe, dvc, PduResult};
+use ironrdp_svc::{AsAny, CompressionCondition, StaticVirtualChannel};
+use pdu::cursor::WriteCursor;
+use pdu::PduEncode;
+
+/// A type that is a Dynamic Virtual Channel (DVC)
+///
+/// Dynamic virtual channels may be created at any point during the RDP session.
+/// The Dynamic Virtual Channel APIs exist to address limitations of Static Virtual Channels:
+///   - Limited number of channels
+///   - Packet reconstruction
+pub trait DynamicVirtualChannel: Send + Sync {
+    fn channel_name(&self) -> &str;
+
+    fn process(&mut self, channel_id: u32, payload: &[u8], output: &mut WriteBuf) -> PduResult<()>;
+}
+
+assert_obj_safe!(DynamicVirtualChannel);
+
+/// DRDYNVC Static Virtual Channel (the Remote Desktop Protocol: Dynamic Virtual Channel Extension)
+///
+/// It adds support for dynamic virtual channels (DVC).
+pub struct Drdynvc {
+    dynamic_channels: BTreeMap<String, Box<dyn DynamicVirtualChannel>>,
+}
+
+impl fmt::Debug for Drdynvc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Drdynvc([")?;
+
+        let mut is_first = true;
+
+        for channel in self.dynamic_channels.values() {
+            if is_first {
+                is_first = false;
+            } else {
+                write!(f, ", ")?;
+            }
+
+            write!(f, "{}", channel.channel_name())?;
+        }
+
+        write!(f, "])")
+    }
+}
+
+impl Drdynvc {
+    pub const NAME: ChannelName = ChannelName::from_static(b"drdynvc\0");
```

Comment:
nit: sometimes this interferes with some of the `PduError` macros who expect a `NAME: &'static str`, maybe we can change it to `CHANNEL_NAME`.

Reply:
I see, yes we could do that! (I’m not sure [this](https://github.com/Devolutions/IronRDP/blob/cd91756305eea95702510e48367754eba5e2ac64/crates/ironrdp-pdu/src/macros.rs#L21) should be used outside of `PduEncode` / `PduDecode` though)

Reply:
@CBenoit however `NAME` note that is used for a short form of `PduError` building macros such as `custom_err!` https://github.com/Devolutions/IronRDP/blob/feat/clipboard-text/crates/ironrdp-pdu/src/macros.rs#L120-L122

Reply:
Maybe this should be de-sugared into `Self::NAME` for all macros so it’s more versatile, but I think there was a problem because of "conflicting" definitions with multiple traits. I have an idea on how to improve that long term.